File: FBSimulatorControl/Commands/FBSimulatorDapServerCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorDapServerCommands.h"
#import "FBSimulator.h"

@interface FBSimulatorDapServerCommand ()

@property (nonatomic, strong, readonly) FBSimulator *simulator;

@end

@implementation FBSimulatorDapServerCommand

#pragma mark Initializers
+ (instancetype)commandsWithTarget:(FBSimulator *)target
{
  return [[self alloc] initWithSimulator:target];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;

  return self;
}

- (FBFuture<FBProcess<id, id<FBDataConsumer>, NSString *> *> *) launchDapServer:dapPath stdIn:(FBProcessInput *)stdIn stdOut:(id<FBDataConsumer>)stdOut{
  NSString *dap_log_dir = [self.simulator.coreSimulatorLogsDirectory stringByAppendingPathComponent:@"dap"];
  
  NSError *error = nil;
  BOOL createdDir = [[NSFileManager defaultManager] createDirectoryAtPath:dap_log_dir
                    withIntermediateDirectories:YES
                    attributes:nil
                    error:&error];
  
  if (!createdDir) {
    return [[FBControlCoreError
             describeFormat:@"Dap Command: Failed to create log director on path %@. Error: %@", dap_log_dir, error.localizedDescription]
      failFuture];
  }
  
  NSString *log_string = [dap_log_dir stringByAppendingPathComponent:[NSUUID.UUID.UUIDString stringByAppendingString:@".log"]];
  BOOL createdLogFile = [[NSFileManager defaultManager] createFileAtPath:log_string
                          contents:nil
                          attributes:nil];
  if (!createdLogFile) {
    return [[FBControlCoreError
      describeFormat:@"Failed to create log file on path %@", log_string]
      failFuture];
  }
  
  [self.simulator.logger.debug logFormat:@"Dap Command: Launching dap server logging at path %@", log_string];
  NSDictionary<NSString *, NSString *> *envs = @{
    @"LLDBVSCODE_LOG": log_string
  };
  NSString *fullPath = [self.simulator.dataDirectory stringByAppendingPathComponent:dapPath];
  return [[[[[[FBProcessBuilder
              withLaunchPath:fullPath]
              withEnvironment:envs]
              withStdIn:stdIn]
              withStdOutConsumer: stdOut]
              withStdErrInMemoryAsString]
              start];
}

@end

File: FBSimulatorControl/Commands/FBSimulatorDapServerCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

@objc(FBSimulatorDapServerCommand)
public final class FBSimulatorDapServerCommand: NSObject, FBiOSTargetCommand {

  // MARK: - Properties

  private let simulator: FBSimulator

  // MARK: - Initializers

  @objc(commandsWithTarget:)
  public class func commands(with target: any FBiOSTarget) -> FBSimulatorDapServerCommand {
    return FBSimulatorDapServerCommand(simulator: target as! FBSimulator)
  }

  private init(simulator: FBSimulator) {
    self.simulator = simulator
    super.init()
  }

  // MARK: - FBDapServerCommand (legacy FBFuture entry point)

  @objc
  public func launchDapServer(_ dapPath: Any, stdIn: FBProcessInput<AnyObject>, stdOut: any FBDataConsumer) -> FBFuture<FBSubprocess<AnyObject, any FBDataConsumer, NSString>> {
    fbFutureFromAsync { [self] in
      try await launchDapServerAsync(dapPath, stdIn: stdIn, stdOut: stdOut)
    }
  }

  // MARK: - Private

  fileprivate func launchDapServerAsync(_ dapPath: Any, stdIn: FBProcessInput<AnyObject>, stdOut: any FBDataConsumer) async throws -> FBSubprocess<AnyObject, any FBDataConsumer, NSString> {
    let dapLogDir = (simulator.coreSimulatorLogsDirectory as NSString).appendingPathComponent("dap")

    do {
      try FileManager.default.createDirectory(atPath: dapLogDir, withIntermediateDirectories: true, attributes: nil)
    } catch {
      throw
        FBControlCoreError
        .describe("Dap Command: Failed to create log director on path \(dapLogDir). Error: \(error.localizedDescription)")
        .build()
    }

    let logString = (dapLogDir as NSString).appendingPathComponent(UUID().uuidString + ".log")
    let createdLogFile = FileManager.default.createFile(atPath: logString, contents: nil, attributes: nil)
    if !createdLogFile {
      throw
        FBControlCoreError
        .describe("Failed to create log file on path \(logString)")
        .build()
    }

    simulator.logger?.debug().log("Dap Command: Launching dap server logging at path \(logString)")
    let envs: [String: String] = [
      "LLDBVSCODE_LOG": logString
    ]
    guard let dataDirectory = simulator.dataDirectory else {
      throw FBControlCoreError.describe("Simulator has no data directory").build()
    }
    let fullPath = (dataDirectory as NSString).appendingPathComponent(dapPath as! String)
    let startedFuture = FBProcessBuilder<AnyObject, AnyObject, NSString>
      .withLaunchPath(fullPath)
      .withEnvironment(envs)
      .withStdIn(stdIn)
      .withStdOutConsumer(stdOut)
      .withStdErrInMemoryAsString()
      .start()
    return try await bridgeFBFuture(startedFuture)
  }
}

// MARK: - AsyncDapServerCommand

extension FBSimulatorDapServerCommand: AsyncDapServerCommand {

  public func launchDapServer(
    _ dapPath: Any,
    stdIn: FBProcessInput<AnyObject>,
    stdOut: any FBDataConsumer
  ) async throws -> FBSubprocess<AnyObject, FBDataConsumer, NSString> {
    try await launchDapServerAsync(dapPath, stdIn: stdIn, stdOut: stdOut)
  }
}

// MARK: - FBSimulator+AsyncDapServerCommand

extension FBSimulator: AsyncDapServerCommand {

  public func launchDapServer(
    _ dapPath: Any,
    stdIn: FBProcessInput<AnyObject>,
    stdOut: any FBDataConsumer
  ) async throws -> FBSubprocess<AnyObject, FBDataConsumer, NSString> {
    try await dapServerCommand().launchDapServerAsync(dapPath, stdIn: stdIn, stdOut: stdOut)
  }
}
